crates/ty_python_semantic/src/semantic_index/builder.rs

```diff
@@ -1507,14 +1507,16 @@ impl<'ast> Visitor<'ast> for SemanticIndexBuilder<'_, 'ast> {
 
                 let pre_loop = self.flow_snapshot();
                 let predicate = self.record_expression_narrowing_constraint(test);
-                self.record_reachability_constraint(predicate);
 
                 // We need multiple copies of the reachability constraint for the while condition,
                 // since we need to model situations where the first evaluation of the condition
                 // returns True, but a later evaluation returns False.
                 let first_predicate_id = self.current_use_def_map_mut().add_predicate(predicate);
+                let first_reachability_constraint = self
+                    .current_reachability_constraints_mut()
+                    .add_atom(first_predicate_id);
```

Comment:
It looks like we can replace all four of these lines, and the below call to `self.record_reachability_constraint_id(first_predicate_id)`, with the single line:
```suggestion
                let first_reachability_constraint = self.record_reachability_constraint(predicate);
```

and avoid some repeat work, too.

Also, the comment line below "So the body's full reachability constraint is `first`." might need a bit of attention, since we no longer have anything here named `first`.